Create global and auth controllers

Peri7at opened the pull request:
Closes #13 and #26

AmmarBaki2 commented on src/controllers/item.js:
```diff
@@ -1,6 +1,13 @@
 module.exports = {
-  getAllItems: (req, res) => {
-    res.send('All items');
+  getAllItems: async (req, res) => {
+    try {
+      const items = await ItemModel.find();
+      res.json(items);
```
When there are no items, `.find()` method doesn't throw an error, instead, it returns an empty array. 
You can check the length of `items` if it is zero or less, then throw an error with message that indicates there are no items

Peri7at replied:
Thank you Ammar, good suggestion, I will revise that part.

AmmarBaki2 commented on src/controllers/user.js:
```diff
@@ -14,7 +16,12 @@ module.exports = {
   },
-  getAllDonators: (req, res) => {
-    res.send('All donators');
+  getAllDonators: async (req, res) => {
+    try {
+      const donators = await UserModel.find({ isDonator: true });
+      res.json(donators);
```
The same issue as in items, you can throw an error here too.

AmmarBaki2 reviewed (changes requested):
Excellent work @Peri7at! 
I added a suggestion below 😁

AmmarBaki2 commented on src/controllers/item.js:
```diff
@@ -4,7 +4,8 @@ module.exports = {
     try {
       const items = await ItemModel.find();
-      res.json(items);
+      if (items.length < 0) res.json({ message: 'No items found' });
+      else res.json(items);
```
To make it easier for front-end developers to differentiate between when there are items and when there are none, throwing an error would be the best practice. Checking the response code would be enough. 
Another thing is that when items length is zero, it will go to the else statement, instead of less than it should be less than or equal to zero.

Peri7at replied:
Thank you @AmmarBaki2 for noting that out. All resolved.

mkkasem replied:
I have a different perspective about returning error when you have an empty array. It's totally fine to return an error as Ammar said, but I think it's better to return either null or the empty array itself because  it's not an error and the frontend developer can notice the difference between an error and empty array.
You can check this stackoverflow question

https://stackoverflow.com/questions/24852784/return-empty-array-or-error/24852870#24852870

Peri7at replied:
It agree with you Mahmoud that it is not an error. Thank you for sharing your perspective. 
Let's see Shrreya's opinion.. @Shrreya what front end developers would like to get for an empty array?

Shrreya replied:
@Peri7at @mkkasem @AmmarBaki2 Interesting discussion!

Firstly, I agree that an empty array is not an error. The server did what it was requested to, there's just no content to return to the client. So it should be a regular 200 OK response code, not an error response code.

But coming to @AmmarBaki2's point about using an error code that helps frontend devs to differentiate, this is usually an internal code, not an HTTP error response code. Or sometimes backend devs return the array length along with the array as extra information in the response body.

I think you can keep it as it for now, frontend devs can check the array length to handle no items scenario.

Peri7at replied:
Thank you for the info @Shrreya. So I will change back to _**if (items.length <= 0) res.json({ message: 'No items found' });**_, right? Right now it is _**if (items.length <= 0) throw new Error('No items found');**_

Shrreya replied:
@Peri7at No change required.
The error thrown will go to catch block and then res.json will be called there.

AmmarBaki2 commented on src/controllers/user.js:
```diff
@@ -19,7 +19,8 @@ module.exports = {
     try {
       const donators = await UserModel.find({ isDonator: true });
-      res.json(donators);
+      if (donators.length < 0) res.json({ message: 'No donators found' });
+      else res.json(donators);
```
Same things here 😅

AmmarBaki2 commented:
Also forgot to mention that you need to merge the main branch with your branch since Ammar merged the bug fix PR.

Shrreya commented:
@Peri7at Please tag me here once you've resolved the changes requested from Ammar's review. Then I'll add my review.

Peri7at commented:
> @Peri7at Please tag me here once you've resolved the changes requested from Ammar's review. Then I'll add my review.

@Shrreya I have pushed last changes after Ammar's review. Please add your review. Thank you.

AmmarBaki2 reviewed (approved):
Excellent job 🥳. Thank you for the changes 😊

Shrreya commented on src/controllers/item.js:
```diff
@@ -1,6 +1,14 @@
+      if (items.length <= 0) throw new Error('No items found');
+      else res.json(items);
+    } catch (err) {
+      res.status(422).json({ message: err.message });
```
You should change this to: `message: err.message ?? err`.

The custom error you are throwing above is just a string, not an object with a message property. But some mongoose internal errors may be objects with message property. So this ensures that the message value is pulled in correctly for the response body.

Shrreya reviewed (changes requested):
Good work @Peri7at!

Other than the change request mentioned below, please note: You have to write tests for all your controllers. So I can approve this PR only after the tests for these 2 controller functions are added.

Peri7at commented:
> Good work @Peri7at!
> 
> Other than the change request mentioned below, please note: You have to write tests for all your controllers. So I can approve this PR only after the tests for these 2 controller functions are added.

@Shrreya I have added the tests for getAllDonators and getAllItems. Please review my PR. Thank you in advance.

AmmarBaki2 commented on src/db/connection.js:
```diff
@@ -32,7 +32,8 @@ const clearDatabase = async () => {
   /* eslint-disable no-restricted-syntax, guard-for-in */
   for (const key in collections) {
-    results.push(mongoose.connection.dropCollection(key));
+    // results.push(mongoose.connection.dropCollection(key));
+    results.push(collections[key].deleteMany());
```
@Shrreya I just want to add a note here, we spent the entire day yesterday trying to solve some error. There was an error saying `MongoError: ns not found` and after a lot of testing I found that it is better to delete what is inside the collection rather than dropping the collection, because this error is occurring due to trying to drop a collection that does not exist,  the test is running fine in the first test file, but when the second test file starts running we are calling this function again so it comes here and try to drop the collections that are already dropped in the previous test (I don't know why but keys are still existing inside the `cellections` variable even though we dropped them).

AmmarBaki2 replied:
and because this error is thrown, the `afterAll` function is not running everything inside it so `server.close()` is not being called and another error is occurring that says `port already in use`
If we leave this function like this, we will get rid of these two errors

Shrreya replied:
@AmmarBaki2 Thank you for letting me know! Looks like you had an intense debugging session, but happy to know you found the issue and resolved it.

It's a good learning that `deleteMany()` does a cleaner job than `dropCollection()`.

Shrreya commented on .eslintrc.json:
```diff
@@ -17,6 +17,7 @@
     "no-process-exit": "off",
-    "class-methods-use-this": "off"
+    "class-methods-use-this": "off",
+    "arrow-body-style": "warn"
```
What was the reason for adding this rule?

Peri7at replied:
Eslint was giving an error for arrow function inside describe and tests were not taking place, so I added, but it was when I was trying to use mockingoose I guess, I don't remember exactly. 
Now I tried to run without this rule and it is working fine, so I removed it.

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,94 @@
+
+    res.clearCookie('token');
   },
 };
```
Was this auth file merged from another branch? Then I won't review it in this PR.

Shrreya replied:
Oh okay, then can you create an issue for setting up auth routes and link it to this PR?
I'll do another review of the auth code as well.

Peri7at replied:
I guess I did it, but I am not sure if I did it correctly

Shrreya replied:
Yes it's done! You can confirm by seeing at the bottom of the PR activities and also on the right hand side it shows that you linked the issue.

Can you update your PR description to "Closes #13 and #26"?

Peri7at replied:
Thank you, Done

Shrreya reviewed (changes requested):
@Peri7at Good job! Just a few small changes requested.

Peri7at commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,88 @@
+      expect(res.header['content-type']).toContain('application/json');
+      expect(res.statusCode).toBe(200);
+      expect(res.body.length).toBe(1);
+      expect(res.body[0].username).toBe('new.user');
```
I created new donator object with field 'isDonator: true' to match the whole object, because the user's default field for isDonator is false.

Shrreya commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,97 @@
+
+let userId;
+
+const newValidUser = {
```
You could rename this to `donatorUser`.

Shrreya commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,97 @@
+  acceptTos: 'on',
+};
+
+const newValidUser2 = {
```
And rename this to `notDonatorUser`.

Shrreya commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,97 @@
+  password: 'testPassword2',
+  password2: 'testPassword2',
+  address: 'new address2',
+  acceptTos: 'on',
```
Is `acceptTos` property used in your authentication logic?

Peri7at replied:
It is implemented in POST /signup.

Shrreya commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,97 @@
+  acceptTos: 'on',
+};
+
+const donator = {
+  username: 'new.user',
+  firstName: 'New',
+  lastName: 'User',
+  email: 'email@domain.com',
+  address: 'new address',
+  isDonator: true,
+};
```
You can rename this to `expectedDonatorResponse`.

Shrreya commented on src/controllers/__tests__/user.test.js:
```diff
@@ -0,0 +1,97 @@
+      // eslint-disable-next-line no-underscore-dangle
+      userId = responseBody[0]._id;
+
+      await request(server)
+        .put(`/api/users/${userId}`)
+        .set('Content-Type', 'application/json')
+        .send({ isDonator: true });
```
Can you add a comment above this code explaining why this endpoint is being called. I know it is to set the user as a donator but always good to add explainer comments like this.

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,94 @@
-  signInUser: (req, res) => {
-    res.send('You successfully signed in');
+  signInUser: async (req, res) => {
+    res.json('');
```
Why is this function commented out?

Peri7at replied:
Because I needed just signup functionality, but since it got assigned to me I will uncomment and work on it.

Shrreya commented on src/controllers/user.js:
```diff
@@ -1,20 +1,53 @@
   },
-  getSingleUser: (req, res) => {
-    res.send('Single user');
+  getSingleUser: async (req, res) => {
+    const { id } = req.params;
+    try {
+      const user = await UserModel.findById(id);
+      if (!user) throw new Error("The user with the specified ID wasn't found");
+      res.json(user);
+    } catch (err) {
+      res.status(422).json({ message: err.message });
+    }
```
Was the main branch merged correctly? Because I think this function was updated in another PR that has been merged to main.

Peri7at replied:
User endpoints are not merged yet. I will delete getSingleUser, but I have to leave getAllUsers and updateUser for my tests to pass.

Shrreya reviewed (changes requested):
@Peri7at Suggested some changes below. And now that the scope of this PR got expanded with auth routes, you will also have to write auth tests. 🙈

Peri7at commented:
Heyy @Shrreya, I have fixed auth routes and added tests for them (#26), also resolved last change requests related to global route controller tests. Please review my PR. Thank you in advance.

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,92 @@
+    const { username, email, password, rememberMe } = req.body;
+    let cookieAge = 24 * 3600;
+    try {
+      const user = await User.findOne({
+        $or: [{ email: email }, { username: username }],
+      });
+      /* eslint object-shorthand: ["warn"] */
```
I think it should work even if you write this as: `$or: [{ email }, { username }]`

Can you check?

Then you won't need `/* eslint object-shorthand: ["warn"] */`

Peri7at replied:
Yes, it works. Thank you.

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,92 @@
+        $or: [{ email: email }, { username: username }],
+      });
+      /* eslint object-shorthand: ["warn"] */
+      if (!user) throw new Error('Wrong username');
```
Shouldn't this message be `Wrong username or email` since both can be used to login?

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,92 @@
+      };
+
+      if (rememberMe) {
+        cookieAge = 14 * 24 * 3600;
```
You can rewrite this and add a comment as: `cookieAge = 14 * cookieAge; // Remember me setting extends cookie expiry time to 2 weeks`

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,92 @@
+  // eslint-disable-next-line consistent-return
+  signUpUser: async (req, res) => {
+    try {
+      const {
+        firstName,
+        lastName,
+        username,
+        email,
+        password,
+        password2,
+        address,
+        acceptTos,
+      } = req.body;
```
It would be nice to rename `password2` to `passwordConfirm`.
And `acceptTos` to `acceptTerms`. However, does this field exist in the user model schema?

Peri7at replied:
`acceptTerms` doesn't exist on user model schema. Should I add it to schema and assign its type to boolean\ default`true` or String/ default 'on', and make it required? But then I need to check if it is `true` or `on` just then make it possible for user to sign up?

Shrreya replied:
Yes add it to the schema as a boolean, but don't put default or required on it.
So that frontend developers can have the option to not use it.

Shrreya commented on src/controllers/auth.js:
```diff
@@ -1,11 +1,92 @@
+      res.clearCookie('_t');
+      res.json({ success: true });
+    } catch (err) {
+      res.status(422).json({ message: err.message });
```
Don't forget `err.message ?? err`
In case any internally thrown error does not have message property.

Peri7at replied:
I thought it doesn't need to be `err.message ?? err`, because I am not throwing any error on try block..

Shrreya replied:
Yes but it's always safe to use it, because even some internal errors thrown by the functions you're calling may or may not have a message property.

Peri7at replied:
Okay, thank you, good to know

Shrreya commented on src/controllers/user.js:
```diff
@@ -1,20 +1,43 @@
 module.exports = {
-  getAllUsers: (req, res) => {
-    res.send('All users');
+  getAllUsers: async (req, res) => {
+    try {
+      const users = await UserModel.find();
+      res.json(users);
+    } catch (err) {
+      res.status(422).json({ message: err.message ?? err });
+    }
```
Can you add a comment above this function and `updateUser`?
It can say: `// Temporary code added, to be merged and updated in upcoming PR`

Shrreya reviewed (changes requested):
@Peri7at Amazing work! I'm so impressed by the clean tests you all are writing.

Here are some final changes requested. Let's aim to get this PR approved and merged tonight.

Peri7at commented on src/controllers/__tests__/auth.test.js:
```diff
@@ -0,0 +1,189 @@
+  describe('POST /api/auth/signup', () => {
+    test('Should sign up a new user', async () => {
+      const res = await request(server).post('/api/auth/signup').send(mockUser);
+      expect(res.statusCode).toBe(200);
```
Will it be enough to run  `expect(res.body).toBeDefined()`?

Shrreya replied:
Since you're testing the full user record on the DB, you can test `toBeDefined()` and any one user property on the response body.

Peri7at replied:
Okay, thanks

Peri7at commented on src/controllers/__tests__/auth.test.js:
```diff
@@ -0,0 +1,189 @@
+      expect(res.text).toMatch(/wrong password/i);
+    });
+
+    test('Should sign in and return token', async () => {
```
Should I rename to `Should return valid token after sign in` or just `Should return valid after sign in`?

Shrreya replied:
Ah sorry I made a typo, it should be `Should return valid token after sign in`

Peri7at commented on src/controllers/__tests__/auth.test.js:
```diff
@@ -0,0 +1,189 @@
+        .send(correctUser1)
+        .set('Authorization', `Bearer ${token}`);
+
+      const eat = new Date();
+      eat.setDate(eat.getDate() + 1);
```
expired at as iat (issued at), shall I rename it to just `date`?

Peri7at replied:
I have changed it to `tokenExpiryDate`

Shrreya replied:
Yup that's better!

Peri7at commented on src/controllers/__tests__/auth.test.js:
```diff
@@ -0,0 +1,189 @@
+      expect(res.header.location).toBe(undefined);
+      expect(res.statusCode).toBe(422);
+      expect(res.text).toMatch(/passwords do not match/i);
+    });
```
We haven't implemented password validation, nor email validation. Shall I use express-validator? Should I implement it in app.js before using auth routes as middleware or directly on auth.js?

Shrreya replied:
Okay then in the interest of getting this PR merged quickly, for now just create an issue for password validation.
Make sure that someone in the team takes up the issue in this week's sprint.
For now we can leave it out of this PR.

Peri7at commented:
> @Peri7at We're almost there! 🏃🏼‍♀️
> Just some last quick changes.

🏃🏼‍♀️🏃🏼‍♀️Done

Shrreya reviewed (approved):
🚀
